aaelf64/aaelf64.rst

```diff
@@ -1249,10 +1249,57 @@ Finally the bit-field of X specified in the table (those bits of X picked out by
 
 For "``MOVW``" type relocations it is the assembler’s responsibility to encode the hw bits (bits 21 and 22) to indicate the bits in the target value that the immediate field represents.
 
-Relocation relaxation
-^^^^^^^^^^^^^^^^^^^^^
+Relocation optimization
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Linkers may optionally optimize instructions affected by relocation. Relocation optimizations improve the efficiency of relocated instructions without changing their visible behaviour. There are several classes of relocation optimizations:
+
+- A single relocation optimization may change an instruction after relocation into an equivalent, more efficient form.
+
+  - Several relocations may result in an addition with zero, which may be optimized as follows:
+
+    ::
+
+      ADD x0, x1, 0   // eg. R_<CLS>_TLSLE_ ADD_TPREL_HI12
+      ADD x2, x2, 0   // or R_<CLS>_ADD_ABS_LO12_NC
+
+      // after optimization:
+
+      MOV x0, x1
+      NOP
+
+  - The relocation ``R_<CLS>_ADR_PREL_PG_HI21`` may emit a MOV with zero immediate for undefined weak symbols.
+
+- A sequence of relocations can be relaxed into more efficient relocations. All of the following conditions must be true:
```

Comment:
"A sequence of relocations can be relaxed into more efficient relocations" reads a bit strangely, and we'll also need to be careful with ``R_<CLS>_ADD_ABS_LO12_NC`` as this will need to do more than just relocate the immediate field, it will also need to change the instruction. As we've not got a relocation code for rewrite this ``ld`` as an ``add`` then the static linker will most likely have to introduce an internal one or rewrite the instruction separately.

I suggest:
A sequence of instructions with relocations, under certain conditions, can be relaxed into more efficient instructions. For the relaxation all of the following conditions must be true:
...

Reply:
Indeed, I'm not suggesting that linkers literally change GOT relocations to PC-relative internally.